### ISSUE TITLE
Integration test for Joomla and WordPress restoration

### ISSUE THREAD

**Issue by nikosdion** [akeeba/angie#42]
Please note that this item is about doing integration (Selenium) testing for ANGIE itself, not Akeeba Backup or Kickstart.

We need to be able to check _automatically_ that backups taken with Akeeba Backup for Joomla and WordPress restore _correctly_. Currently our automated tests only cover that the restoration script can execute but don't check the restoration flow or the results.

This is a lot of work but it would help us ensure that all sorts of WordPress and Joomla weirdness is dealt with by ANGIE.

> *IMPORTANT FORWARD COMPATIBILITY NOTE*: Always reference items by their ID. Do not assume that the DOM structure in ANGIE pages will be immutable. When we convert it to FEF the DOM will change. If you need to check if a message is present look for the raw text in the page output. Get the raw text _from the language string_, do NOT hard-code it (messages might also change in the future).

### Technologies used

We can use PHPunit as the execution engine with Facebook Selenium WebDriver package to interface Selenium. On the test machine Selenium WebDriver standalone controlling Firefox headless will be running the tests. This is the same setup we use for Akeeba Backup and Akeeba Solo integration testing. It's reliable and we have experience on it.

### Approach and configuration

We only need to test MySQL restoration. PostgreSQL and MS SQL Server are cumulatively used by less than 0.3% of our clients.

We only test with the lowest supported PHP version, currently PHP 5.3. The reason is that servers tend to assign their oldest PHP version as the default in the absence of a .htaccess, i.e. when we are restoring (sigh).

Our tests will create model Joomla and WordPress sites with predictable sample content. The user need to provide us with the disk locations, URLs and db information required to create those sites. For the WordPress sites look into [my work on making test sites](https://github.com/nikosdion/wpsites). You can use a similar approach for Joomla. Caveat: Joomla does not have CLI method to create content so you'll probably have to provide SQL files instead, the same way J's sample content works :/

Then we install Akeeba Backup and have it take backups using the CLI scripts. The path to Akeeba Backup Professional packages should be given in the configuration.

The disk location, URLs and DB connection info for restoration is also given in the configuration. We extract each backup to its relevant directory.

On each target folder we only keep the installation/sql folder and the generated INI and HTML files in the installation folder but remove all other ANGIE files. Then we rename `installation` to `installation_source`.

Everything until this point we should be able to run at our discretion, NOT every time we run the tests. It should be a separate script which sets up our environment.

Each test creates the `installation` folder afresh, copies the relevant ANGIE installer inside it and then copies the contents of `installation_source` into the `installation` folder. This gives us a fresh installer which is up to date with any changes we have made to ANGIE e.g. if we spot bugs and are trying to fix them.

### Joomla restoration tests

1. Backup taken on the domain root (e.g. `http://joomla.local.web`) restored onto itself. This is required to check that the database data is NOT removed automatically.
1. Backup taken on the domain root (e.g. `http://joomla.local.web`) restored to a different domain's root (e.g. `http://restored.local.web`).
1. Backup taken on the domain root (e.g. `http://joomla.local.web`) restored to a subdirectory (e.g. `http://restored.local.web/joomla`)
1. Backup taken on a subdirectory (e.g. `http://joomla.local.web/joomla`) restored to a different subdirectory (e.g. `http://restored.local.web/foobar`)
These tests need to make sure the resulting site works, images and CSS is loaded without 404s.

Test #1 should also include variations for:
* Using ANGIE password
* Restoring with the database data blanked out (databases.ini does not contain db information)
* Restoring multiple databases
* Deleting .htaccess, .htpasswd, php.ini and .user.ini files
* Check that modifying the Site Setup information does have the intended effect in configuration.php
These subtests do not need to check if the resulting site actually works.

### WordPress restoration tests

I am referencing the names of the sample sites from my wpsites repository to figure it out easier.

1. http://www.regular.wp/ to http://restored.wp
1. http://www.regular.wp/ to http://restored.wp/subdirectory
1. http://www.subdir.wp/blog/ to http://restored.wp
1. http://www.subdir.wp/blog/ to http://restored.wp/subdirectory
1. http://www.rewritten.wp/ to http://restored.wp
1. http://www.rewritten.wp/ to http://restored.wp/subdirectory
1. http://multi.wp to http://restored.wp, using subdomains on the restored site for multisites (check the subdomains blog1.restored.wp and blog2.restored.wp) - This is PRESERVING subdomain multisites
1. http://multi.wp to http://restored.wp, using subfolders on the restored site for multisites (check the subfolders restored.wp/blog1 and restored.wp/blog2) - This is FROM subdomain TO subfolder multistes
1. http://www.folders.wp to http://restored.wp, using subdomains on the restored site for multisites (check the subdomains blog1.restored.wp and blog2.restored.wp) - This is PRESERVING subfolder multisites
1. http://www.folders.wp to http://restored.wp, using subfolders on the restored site for multisites (check the subfolders restored.wp/blog1 and restored.wp/blog2) - This is FROM subfolder TO sudomain multisites
1. http://www.multifolder.wp/blog to http://restored.wp/subfolder, using subfolders on the restored site for multisites (check the subfolders restored.wp/subfolder/blog1 and restored.wp/subfolder/blog2) - This is FROM subfolder TO sudomain multisites WITH a root subfolder on the source
1. http://www.multifolder.wp/blog to http://restored.wp, using subdomains on the restored site for multisites - This is FROM subfolder TO sudomain multisites WITH a root subfolder on the source
1. http://multicustom.wp with the mapping multicustom.wp -> restored.wp, www.omg.wp -> restored2.wp and moar.wp -> restored3.wp. In this case multicustom/restored use subdomains, omg.wp/restored2.wp is like a standalone site (technically: uses subdomain multisite with a custom domain) and moar/restored3 use subfolder multisites. No, remapping this lunacy to use a different multisite setup is out of scope and the premise of a madman.

**Comment by nikosdion** [comment]
The main problem we had was WordPress data replacement. Moreover, the approach outlined here was very slow and finicky because it involved PHPUnit, WebDriver and a Vagrant box. Every time one of them was updated we had to rewrite the tests even if our software didn't change. This is a waste of resources. Tests should save us time and effort, not create another time sink.

For this reason I wrote a more efficient suite of tests in its own separate repository.  It uses UNiTE (which uses ANGIE for the restoration) and a simple list of URLs expected to be present on the page AND load without HTTP errors to check whether the restoration completed successfully.

This ticket will be closed after the removal of the Tests folder.
